src/EventStore.Core/Services/Replication/TransactionFramer.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Collections.Generic;
+using EventStore.Common.Utils;
+using EventStore.Core.TransactionLog.LogRecords;
+using EventStore.Transport.Tcp.Framing;
+
+namespace EventStore.Core.Services.Replication; 
+
+public class TransactionFramer : IMessageFramer<ILogRecord[]> {
+	private readonly List<ILogRecord> _records;
+	private readonly IMessageFramer<ILogRecord> _inner;
+	private Action<ILogRecord[]> _handler;
+
+	public TransactionFramer(IMessageFramer<ILogRecord> inner) {
+		_records = new List<ILogRecord>(capacity: 1024);
+		_inner = inner;
+		_inner.RegisterMessageArrivedCallback(OnLogRecordUnframed);
+	}
+
+	public bool HasData => _records.Count > 0 || _inner.HasData;
+	public IEnumerable<ArraySegment<byte>> FrameData(ArraySegment<byte> data) => _inner.FrameData(data);
+	public void UnFrameData(IEnumerable<ArraySegment<byte>> data) => _inner.UnFrameData(data);
+	public void UnFrameData(ArraySegment<byte> data) => _inner.UnFrameData(data);
+
+	public void Reset() {
+		_records.Clear();
+		_inner.Reset();
+	}
+
+	private void OnLogRecordUnframed(ILogRecord record) {
+		_records.Add(record);
+		if (record.IsTransactionBoundary()) {
+			_handler?.Invoke(_records.ToArray());
```

Comment:
are the ToArray() calls here and below important? it's quite a lot of allocations we could possibly avoid

Reply:
well spotted, fixed :+1: